x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CacheFileTests.java

```diff
@@ -171,6 +196,151 @@ public void testConcurrentAccess() throws Exception {
+                            channel -> Math.toIntExact(range.v2() - range.v1()),
+                            (channel, from, to, progressUpdater) -> progressUpdater.accept(to),
+                            threadPool.generic()
+                        );
+
+                        waitForGenericThreadPool(threadPool, expectedCompletedRanges.size() + 1L);
+                        assertTrue(expectedCompletedRanges.add(range));
+                        assertFalse(cacheFile.isFSynced());
+                    }
+
+                    completedRanges = cacheFile.fsync();
+                    assertNumberOfFSyncs(cacheFile.getFile(), equalTo(1L + expectedCompletedRanges.size()));
+                    assertThat(completedRanges.size(), equalTo(expectedCompletedRanges.size()));
+                    assertTrue(cacheFile.isFSynced());
+                }
+
+                assertArrayEquals(cacheFile.fsync().toArray(Tuple[]::new), expectedCompletedRanges.toArray(Tuple[]::new));
+            } finally {
+                cacheFile.release(listener);
+                terminate(threadPool);
+            }
+        }
+    }
+
+    public void testFSyncOnEvictedFile() throws Exception {
+        try (FSyncTrackingFileSystemProvider fileSystem = setupFSyncCountingFileSystem()) {
+            final CacheFile cacheFile = new CacheFile("test", randomNonNegativeLong(), fileSystem.resolve("test"));
+            assertFalse(cacheFile.isFSynced());
+
+            final TestEvictionListener listener = new TestEvictionListener();
+            cacheFile.acquire(listener);
+
+            final TestThreadPool threadPool = new TestThreadPool(getTestName());
+            try {
+                final boolean completeRangeBeforeEviction = randomBoolean();
+                if (completeRangeBeforeEviction) {
+                    final long start = randomLongBetween(0L, Math.max(0L, cacheFile.getLength() - 1L));
+                    final long end = randomLongBetween(start, cacheFile.getLength());
+                    final Tuple<Long, Long> range = Tuple.tuple(start, end);
+                    cacheFile.populateAndRead(
+                        range,
+                        range,
+                        channel -> Math.toIntExact(end - start),
+                        (channel, from, to, progressUpdater) -> progressUpdater.accept(to),
+                        threadPool.generic()
+                    );
+
+                    waitForGenericThreadPool(threadPool, 1L);
+                    assertFalse(cacheFile.isFSynced());
+
+                    final List<Tuple<Long, Long>> completedRanges = cacheFile.fsync();
+                    assertNumberOfFSyncs(cacheFile.getFile(), equalTo(1L));
+                    assertThat(completedRanges, hasSize(1));
+                    assertTrue(cacheFile.isFSynced());
+                }
+
+                cacheFile.startEviction();
+                expectThrows(AlreadyClosedException.class, cacheFile::fsync);
```

Comment:
This can definitely be handled in a follow-up instead when you start using it. But it might be more appropriate to just return the empty set of ranges in this case, since we essentially having nothing cached for it, rather than have to exception handle in the fsync'er thread?

Reply:
I agree. In case of an evicted file the method will now return an empty collection. See https://github.com/elastic/elasticsearch/pull/64201#discussion_r515932191